py/objstr.c

```diff
@@ -2067,6 +2067,8 @@ mp_obj_t mp_obj_new_str_from_vstr(const mp_obj_type_t *type, vstr_t *vstr) {
             vstr->alloc = 0;
             return MP_OBJ_NEW_QSTR(q);
         }
+    } else {
+        type = &mp_type_bytes;
```

Comment:
@dpgeorge the following line was failing:
```
mp_obj_new_str_from_vstr(mp_obj_get_type(where), &vstr_return);
```
when `where` is bytearray etc as this function would copy the incoming type to the output. 
The solution proposed here (in separate commit) is to sanitise the provided type to be either `mp_type_str` or `mp_type_bytes` the same as is done in `mp_obj_new_str_of_type()` earlier in this file.

If this seems like a risky / unneccesary change this type checking could be done as-needed in `modre.c` before calling `mp_obj_new_str_from_vstr`.
However, `mp_type_bytes` is currently not exposed in nativeglue.c so dynamic ure module currently fails with this approach - we could certainly expose `mp_type_bytes` in native modules though if this is a preferred direction; indeed `mp_type_str` is already exposed there.

Reply:
I think what you did here is a good change, I think all existing callers of this function passed either str or bytes, so functionality shouldn't change.  And this scales well to other uses which would have similar semantics.

But actually dynamic ure doesn't enable URE_SUB because dynamic doesn't support `mp_obj_new_str_from_vstr()`!

Reply:
Oh... before this one of the unix builds here was failing, I thought it was from a dynamic build of this.

Reply:
I believe this change is now incorrect _and_ no longer needed after the work done since on standardising str/byte/bytearray functions. As such I've removed it in the rebase, all tests are passing.